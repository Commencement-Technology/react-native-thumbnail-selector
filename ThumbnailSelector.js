import React, {Component, PropTypes} from 'react';
import {
  StyleSheet,
  Text,
  ListView,
  FlatList,
  Image,
  TouchableOpacity,
  Animated,
  Dimensions,
  Modal,
  View,
} from 'react-native';
const {height, width} = Dimensions.get('window');
const WINDOW = Dimensions.get('window');
var timeOutId;

export default class ThumbnailSelector extends Component {
  static propTypes = {
    items: PropTypes.array.isRequired,
    opacity: PropTypes.number,
    fontSize: PropTypes.number,
    textColor: PropTypes.string,
    backgroundColor: PropTypes.string,
    fontFamily: PropTypes.string,
    imageHeight: PropTypes.number,
    imageWidth: PropTypes.number,
    imageBorderWidth: PropTypes.number,
    imageBorderRadius: PropTypes.number,
    onSelectedData: PropTypes.func,
    closeOnSelect: PropTypes.bool,
    zIndex: PropTypes.number,
    closeOnSelectInterval: PropTypes.number,
    numberOfLines: PropTypes.number,
    visible: PropTypes.bool,
  };
  static defaultProps = {
    items: null,
    fontSize: 16,
    textColor: 'white',
    opacity: 0.8,
    backgroundColor: 'dimgray',
    fontFamily: 'Avenir',
    imageHeight: 125,
    imageWidth: 125,
    imageBorderWidth: 2,
    imageBorderRadius: 2,
    onSelectedData: null,
    closeOnSelect: true,
    zIndex: 1000,
    closeOnSelectInterval: 200,
    numberOfLines: 2,
    visible: true,
  };
  constructor(props) {
    super(props);
    this.state = {
      fadeAnim: new Animated.Value(0),
      dataSource: new ListView.DataSource({
        rowHasChanged: (r1, r2) => r1 !== r2,
      }),
      items: props.items,
      visible: false,
      startDelta: WINDOW.height + 155,
      endDelta: 0,
      duration: 600,
    };
    this.show = this.show.bind(this);
    this.hide = this.hide.bind(this);
    this.animateToValue = this.animateToValue.bind(this);
    this.timeOutForisVisible = this.timeOutForisVisible.bind(this);
  }
  componentWillMount() {
    this.setState({
      dataSource: this.state.dataSource.cloneWithRows(this.props.items),
      items: this.props.items,
    });
  }
  componentDidMount() {
    const {visible} = this.props
    if (visible) {
      this.show()
    }
  }
  componentWillReceiveProps(nextProps) {
    const {visible} = this.props
    if (nextProps.visible !== visible) {
      if (nextProps.visible) {
        this.show()
      } else {
        this.hide()
      }
    }
  }
  show() {
    this.animateToValue(1);
  }
  hide() {
    this.timeOutForisVisible(false);
    this.animateToValue(0);
  }
  animateToValue(toValue) {
    Animated.timing(this.state.fadeAnim, {
      toValue: toValue,
      duration: this.state.duration,
    }).start();
  }
  timeOutForisVisible(visible) {
    timeOutId = setTimeout(
      function() {
        this.setState({
          visible: visible,
        });
      }.bind(this),
      this.state.duration,
    );
  }
  onLayoutEvent(event) {
    const {x, y, width, height} = event.nativeEvent.layout;
    this.setState({
      startDelta: WINDOW.height + height,
      endDelta: 0,
    });
  }
  itemAction = (item, index) => {
    var items = this.state.items;
    for (const value of items) {
      value.selected = false;
    }
    item.selected = true;
    this.setState({
      items: items,
    });
    if (this.props.onSelectedItem) {
      this.props.onSelectedItem(item);
    }
    if (this.props.closeOnSelect == true) {
      setTimeout(
        function() {
          this.hide();
        }.bind(this),
        this.props.closeOnSelectInterval,
      );
    }
  };
  renderItem = item => {
    const index = item.index;
    item = item.item;
    return (
      <TouchableOpacity onPress={() => this.itemAction(item, index)}>
        <View
          style={[
            styles.column,
            {opacity: item.selected ? 1 : this.props.opacity},
          ]}
        >
          <Image
            style={{
              borderColor: item.selected ? item.borderColor : 'transparent',
              width: this.props.imageWidth,
              height: this.props.imageHeight,
              borderRadius: this.props.imageBorderRadius,
              borderWidth: this.props.imageBorderWidth,
            }}
            source={{uri: item.imageUri}}
          />
          <Text
            style={{
              color: this.props.textColor,
              fontSize: this.props.fontSize,
              fontFamily: this.props.fontFamily,
              fontWeight: item.selected ? 'bold' : 'normal',
              maxWidth: this.props.imageWidth,
              textAlign: 'center',
            }}
            numberOfLines={this.props.numberOfLines}
          >
            {item.title}
          </Text>
        </View>
      </TouchableOpacity>
    );
  };
  render() {
    const {
      startDelta,
      endDelta,
      fadeAnim,
      dataSource,
      items,
    } = this.state;
    const {zIndex, backgroundColor} = this.props;
    return (
      <Animated.View
        style={{
          zIndex: zIndex,
          transform: [
            {
              translateY: fadeAnim.interpolate({
                inputRange: [0, 1],
                outputRange: [startDelta, endDelta],
              }),
            },
          ],
          position: 'absolute',
          bottom: 0,
        }}
      >
        <FlatList
          ref={ref => this.flatList = ref}
          style={{backgroundColor: backgroundColor, flex: 1}}
          data={items}
          horizontal={true}
          renderItem={this.renderItem}
        />
      </Animated.View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  column: {
    flexDirection: 'column',
    paddingLeft: 8,
    paddingRight: 8,
    paddingTop: 8,
    alignItems: 'center',
  },
});
