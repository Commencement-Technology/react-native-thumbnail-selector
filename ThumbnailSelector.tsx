import React, {useRef, useState} from 'react';
import {
  Text,
  FlatList,
  Image,
  TouchableOpacity,
  Animated,
  useWindowDimensions,
  ImageProps,
  ImageSourcePropType,
  TextProps,
  TouchableOpacityProps,
  ViewStyle,
} from 'react-native';

export type ThumbnailItem = {
  caption?: string;
  imageSrc: ImageSourcePropType;
};

export type ThumbnailItemIndex = {
  item: ThumbnailItem;
  index: number;
};

export type ThumbnailSelectorProps = {
  thumbnails: ThumbnailItem[];
  toggle?: (func: () => void) => void;
  renderThumbnail?: (
    item: ThumbnailItem,
    index: number,
    onSelect?: (item: ThumbnailItem, index: number) => void,
  ) => JSX.Element;
  onSelect?: (item: ThumbnailItem, index: number) => void;
  initialIndex?: number;
  captionCharacterMaxLimit?: number;
  captionEllipsis?: string;
  activeColor?: string;
  inactiveColor?: string;
  imageProps?: ImageProps;
  textProps?: TextProps;
  touchableOpacityProps?: TouchableOpacityProps;
  animatedViewStyle?: ViewStyle;
  animationConfig?: Animated.SpringAnimationConfig;
  flatListViewStyle?: ViewStyle;
};

const ThumbnailSelector: React.FC<ThumbnailSelectorProps> = ({
  thumbnails = [],
  toggle = undefined,
  renderThumbnail = undefined,
  onSelect = undefined,
  initialIndex = -1,
  captionCharacterMaxLimit = 15,
  captionEllipsis = '...',
  activeColor = 'white',
  inactiveColor = 'black',
  imageProps = {
    style: {width: 136, height: 136, borderWidth: 1},
  },
  textProps = {
    style: {fontSize: 16, textAlign: 'center', fontWeight: 'bold'},
    numberOfLines: 1,
  },
  touchableOpacityProps = {
    style: {padding: 8},
  },
  animatedViewStyle = {
    elevation: 1,
    zIndex: 1,
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
  },
  animationConfig = {
    toValue: 0,
    friction: 9,
    useNativeDriver: false,
  },
  flatListViewStyle = {backgroundColor: 'grey', padding: 8},
}) => {
  const window = useWindowDimensions();
  const [itemIndex, setItemIndex] = useState(initialIndex);
  const [animViewHeight, setAnimViewHeight] = useState(0);
  const animatedValue = useRef(new Animated.Value(0));
  const toValue = useRef(animationConfig.toValue);

  const _toggle = () => {
    if (toValue.current) {
      animationConfig.toValue = 0;
    } else {
      animationConfig.toValue = 1;
    }
    toValue.current = animationConfig.toValue;
    Animated.spring(animatedValue.current, animationConfig).start();
  };

  if (toggle) {
    toggle(_toggle);
  }

  const _renderItem = (obj: ThumbnailItemIndex) => {
    const {item, index} = obj;
    if (renderThumbnail) {
      return renderThumbnail(item, index, onSelect);
    }
    const selected = itemIndex === index;
    const color = selected ? activeColor : inactiveColor;
    let caption = item.caption;
    if (caption) {
      if (caption.length > captionCharacterMaxLimit) {
        const end = captionCharacterMaxLimit - captionEllipsis.length;
        caption = `${caption.substring(0, end)}${captionEllipsis}`;
      }
    }
    return (
      <TouchableOpacity
        {...touchableOpacityProps}
        onPress={() => {
          setItemIndex(index);
          if (onSelect) {
            onSelect(item, index);
          }
        }}>
        <Image
          {...imageProps}
          style={[{borderColor: color}, imageProps.style]}
          source={item.imageSrc}
        />
        {caption && (
          <Text {...textProps} style={[{color}, textProps.style]}>
            {caption}
          </Text>
        )}
      </TouchableOpacity>
    );
  };

  const _onLayout = ({nativeEvent = {layout: {height: 0}}}) => {
    const {height} = nativeEvent.layout;
    if (animViewHeight !== height) {
      setAnimViewHeight(height);
    }
  };

  const _getAnimViewStyle = () => {
    const start = window.height;
    const end = window.height - animViewHeight;
    return [
      animatedViewStyle,
      {
        transform: [
          {
            translateY: animatedValue.current.interpolate({
              inputRange: [0, 1],
              outputRange: [start, end],
            }),
          },
        ],
      },
    ];
  };

  return (
    <Animated.View style={_getAnimViewStyle()} onLayout={_onLayout}>
      <FlatList
        style={flatListViewStyle}
        data={thumbnails}
        initialNumToRender={thumbnails.length}
        extraData={itemIndex}
        horizontal={true}
        renderItem={_renderItem}
        keyExtractor={(_item, index) => `${index}`}
      />
    </Animated.View>
  );
};

export default ThumbnailSelector;
