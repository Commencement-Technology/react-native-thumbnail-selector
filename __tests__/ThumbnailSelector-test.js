import 'react-native';
import React from 'react';
import ThumbnailSelector from '../ThumbnailSelector';
import renderer from 'react-test-renderer';

describe('ThumbnailSelector', () => {
  test('open', () => {
    let thumbRef;
    let tree = renderer.create(
      <ThumbnailSelector
        ref={(ref) => {
          thumbRef = ref;
        }}
      />,
    );
    thumbRef.animate();
    expect(tree.toJSON()).toMatchSnapshot();
  });
  test('open then close', () => {
    let thumbRef;
    let tree = renderer.create(
      <ThumbnailSelector
        ref={(ref) => {
          thumbRef = ref;
        }}
      />,
    );
    thumbRef.animate(); // open
    thumbRef.animate(); // close
    expect(tree.toJSON()).toMatchSnapshot();
  });
  test('renders', () => {
    const tree = renderer.create(<ThumbnailSelector />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
