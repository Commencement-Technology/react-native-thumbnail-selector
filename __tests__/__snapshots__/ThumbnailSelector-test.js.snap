// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThumbnailSelector open 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "elevation": 1,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "transform": Array [
        Object {
          "translateY": 1334,
        },
      ],
      "zIndex": 1,
    }
  }
>
  <RCTScrollView
    data={Array []}
    extraData={-1}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    initialNumToRender={0}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</View>
`;

exports[`ThumbnailSelector open then close 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "elevation": 1,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "transform": Array [
        Object {
          "translateY": 1334,
        },
      ],
      "zIndex": 1,
    }
  }
>
  <RCTScrollView
    data={Array []}
    extraData={-1}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    initialNumToRender={0}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</View>
`;

exports[`ThumbnailSelector renders 1`] = `
<View
  collapsable={false}
  nativeID="animatedComponent"
  onLayout={[Function]}
  style={
    Object {
      "elevation": 1,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "transform": Array [
        Object {
          "translateY": 1334,
        },
      ],
      "zIndex": 1,
    }
  }
>
  <RCTScrollView
    data={Array []}
    extraData={-1}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    initialNumToRender={0}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</View>
`;
